satellite/gracefulexit/endpoint.go

```diff
@@ -950,36 +838,105 @@ func (endpoint *Endpoint) updatePointer(ctx context.Context, originalPointer *pb
+		mon.IntVal("graceful_exit_init_node_audit_success_count").Observe(node.Reputation.AuditSuccessCount) //locked
+		mon.IntVal("graceful_exit_init_node_audit_total_count").Observe(node.Reputation.AuditCount)          //locked
+		mon.IntVal("graceful_exit_init_node_piece_count").Observe(node.PieceCount)                           //locked
+
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
 	}
 
-	return false, nil
+	if exitStatus.ExitLoopCompletedAt == nil {
+		return &pb.SatelliteMessage{Message: &pb.SatelliteMessage_NotReady{NotReady: &pb.NotReady{}}}, nil
+	}
+
+	return nil, nil
+}
+
+func (endpoint *Endpoint) generateExitStatusRequest(ctx context.Context, nodeID storj.NodeID) (*overlay.ExitStatusRequest, pb.ExitFailed_Reason, error) {
+	var exitFailedReason pb.ExitFailed_Reason = -1
+	progress, err := endpoint.db.GetProgress(ctx, nodeID)
+	if err != nil {
+		return nil, exitFailedReason, rpcstatus.Error(rpcstatus.Internal, err.Error())
+	}
+
+	mon.IntVal("graceful_exit_final_pieces_failed").Observe(progress.PiecesFailed)         //locked
+	mon.IntVal("graceful_exit_final_pieces_succeess").Observe(progress.PiecesTransferred)  //locked
+	mon.IntVal("graceful_exit_final_bytes_transferred").Observe(progress.BytesTransferred) //locked
+	processed := progress.PiecesFailed + progress.PiecesTransferred
+
+	if processed > 0 {
+		mon.IntVal("graceful_exit_successful_pieces_transfer_ratio").Observe(progress.PiecesTransferred / processed) //locked
+	}
+
+	exitStatusRequest := &overlay.ExitStatusRequest{
+		NodeID:         progress.NodeID,
+		ExitFinishedAt: time.Now().UTC(),
+	}
+	// check node's exiting progress to see if it has failed passed max failure threshold
+	if processed > 0 && float64(progress.PiecesFailed)/float64(processed)*100 >= float64(endpoint.config.OverallMaxFailuresPercentage) {
+		exitStatusRequest.ExitSuccess = false
+		exitFailedReason = pb.ExitFailed_OVERALL_FAILURE_PERCENTAGE_EXCEEDED
+	} else {
+		exitStatusRequest.ExitSuccess = true
+	}
+
+	if exitStatusRequest.ExitSuccess {
+		mon.Meter("graceful_exit_success").Mark(1) //locked
+	} else {
+		mon.Meter("graceful_exit_fail_max_failures_percentage").Mark(1) //locked
+	}
+
+	return exitStatusRequest, exitFailedReason, nil
+
+}
+
+func (endpoint *Endpoint) calculatePieceSize(ctx context.Context, pointer *pb.Pointer, incomplete *TransferQueueItem, nodePiece *pb.RemotePiece) (int64, error) {
+	nodeID := incomplete.NodeID
+
+	// calculate piece size
+	redundancy, err := eestream.NewRedundancyStrategyFromProto(pointer.GetRemote().GetRedundancy())
+	if err != nil {
+		return 0, Error.Wrap(err)
+	}
+
+	pieces := pointer.GetRemote().GetRemotePieces()
+	if len(pieces) > redundancy.OptimalThreshold() {
+		endpoint.log.Debug("pointer has more pieces than required. removing node from pointer.", zap.Stringer("node ID", nodeID), zap.ByteString("path", incomplete.Path), zap.Int32("piece num", incomplete.PieceNum))
```

Comment:
nit - maybe we can move this log statement to where the node is actually removed from the pointer